cpp/src/arrow/extension/fixed_shape_tensor_test.cc

```diff
@@ -0,0 +1,188 @@
+
+auto RoundtripBatch = [](const std::shared_ptr<RecordBatch>& batch,
+                         std::shared_ptr<RecordBatch>* out) {
+  ASSERT_OK_AND_ASSIGN(auto out_stream, io::BufferOutputStream::Create());
+  ASSERT_OK(ipc::WriteRecordBatchStream({batch}, ipc::IpcWriteOptions::Defaults(),
+                                        out_stream.get()));
+
+  ASSERT_OK_AND_ASSIGN(auto complete_ipc_stream, out_stream->Finish());
+
+  io::BufferReader reader(complete_ipc_stream);
+  std::shared_ptr<RecordBatchReader> batch_reader;
+  ASSERT_OK_AND_ASSIGN(batch_reader, ipc::RecordBatchStreamReader::Open(&reader));
+  ASSERT_OK(batch_reader->ReadNext(out));
+};
+
+TEST_F(TestExtensionType, FixedShapeTensorType) {
+  using FixedShapeTensorType = extension::FixedShapeTensorType;
+  using arrow::extension::fixed_shape_tensor;
+
+  std::vector<int64_t> shape = {3, 3, 4};
+  std::vector<int64_t> cell_shape = {3, 4};
+  auto value_type = int64();
+  std::shared_ptr<DataType> cell_type = fixed_size_list(value_type, 12);
+
+  std::vector<std::string> dim_names = {"x", "y"};
+  std::vector<int64_t> strides = {96, 32, 8};
+  std::vector<int64_t> column_major_strides = {8, 24, 72};
+  std::vector<int64_t> neither_major_strides = {96, 8, 32};
+  std::vector<int64_t> cell_strides = {32, 8};
+  std::vector<int64_t> values = {0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11,
+                                 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23,
+                                 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35};
+  std::vector<int64_t> values_partial = {0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11,
+                                         12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23};
+  std::vector<int64_t> shape_partial = {2, 3, 4};
+  std::string serialized = R"({"shape":[3,4],"dim_names":["x","y"]})";
+
+  ASSERT_OK_AND_ASSIGN(auto tensor,
+                       Tensor::Make(value_type, Buffer::Wrap(values), shape));
+  ASSERT_OK_AND_ASSIGN(
+      auto tensor_partial,
+      Tensor::Make(value_type, Buffer::Wrap(values_partial), shape_partial));
+
+  std::shared_ptr<ExtensionType> ext_type =
+      fixed_shape_tensor(value_type, cell_shape, {}, dim_names);
+  auto exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type);
+  ASSERT_OK_AND_ASSIGN(auto ds,
+                       ext_type->Deserialize(ext_type->storage_type(), serialized));
+  std::shared_ptr<ExtensionType> deserialized =
+      std::reinterpret_pointer_cast<ExtensionType>(ds);
+
+  ASSERT_TRUE(tensor->is_row_major());
+  ASSERT_EQ(tensor->strides(), strides);
+  ASSERT_EQ(tensor_partial->strides(), strides);
+
+  // Test ExtensionType methods
+  ASSERT_EQ(ext_type->extension_name(), "arrow.fixed_shape_tensor");
+  ASSERT_TRUE(ext_type->ExtensionEquals(*exact_ext_type));
+  ASSERT_TRUE(ext_type->storage_type()->Equals(*cell_type));
+  ASSERT_EQ(ext_type->Serialize(), serialized);
+  ASSERT_TRUE(deserialized->ExtensionEquals(*ext_type));
```

Comment:
```suggestion
  ASSERT_TRUE(deserialized->Equals(*ext_type));
```

I _think_ public usage should prefer the generic Equals, so good to use that in the tests. And can you also add a case where Equals returns False for comparing with a different type?

Reply:
`ExtensionType` has virtual `ExtensionEquals` but perhaps that should be changed?
I've added an "alias" for now:
`bool Equals(const ExtensionType& other) const { return ExtensionEquals(other); };`

Reply:
That shouldn't be needed. The base class DataType::Equals should already dispatch the ExtensionEquals

Reply:
Maybe it does, but building with `Equals` only throws:
```cpp
/usr/include/c++/11/ext/new_allocator.h:162:11: error: invalid new-expression of abstract class type ‘arrow::extension::FixedShapeTensorType’
  162 |         { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/rok/Documents/repos/arrow/cpp/src/arrow/extension/fixed_shape_tensor.cc:18:
/home/rok/Documents/repos/arrow/cpp/src/arrow/extension/fixed_shape_tensor.h:32:20: note:   because the following virtual functions are pure within ‘arrow::extension::FixedShapeTensorType’:
   32 | class ARROW_EXPORT FixedShapeTensorType : public ExtensionType {
```

Reply:
> Maybe it does, but building with `Equals` only throws:

Does this work now? Since you seem to be using it in the test?

Reply:
Yes, we define `ExtensionEquals` and it gets used in tests by `Equals`.